node_agent/na/nodeagent.go

```diff
@@ -92,13 +101,48 @@ func (na *nodeAgentInternal) Start() {
 		glog.Fatalf("Node Agent is not running on the right platform")
 	}
 
+	// If gracePeriodPercentage is not within [1, 99], we should apply the default (50).
+	gracePeriodPercentage := *na.config.CertRenewalGracePeriodPercentage
+	if gracePeriodPercentage >= 100 || gracePeriodPercentage <= 0 {
+		glog.Errorf("Certificate grace period config %d exceeds limit [1, 99]. Set to default: %d",
+			gracePeriodPercentage, CertRenewalGracePeriodMaxPercentage)
+		gracePeriodPercentage = CertRenewalGracePeriodMaxPercentage
+	}
+
+	retries := 0
+	retrialInterval := CertRequestRetrialInterval
 	for {
+		glog.Info("Sending CSR (retrial #%d) ...", retries)
 		privKey, resp, err := na.sendCSR()
-		if err != nil {
-			glog.Errorf("CSR signing failed: %s", err)
-		} else if resp != nil && resp.IsApproved {
-			timer := time.NewTimer(na.getExpTime(resp))
+		if err == nil && resp != nil && resp.IsApproved {
+			cert, certErr := pki.ParsePemEncodedCertificate(resp.SignedCertChain)
+			if certErr != nil {
+				glog.Fatalf("Error getting TTL from approved cert: %s", certErr)
+			}
+			certTTL := cert.NotAfter.Sub(cert.NotBefore)
+			// Wait until the grace period starts.
+			waittime := certTTL - time.Duration(gracePeriodPercentage/100)*certTTL
+			timer := time.NewTimer(waittime)
 			na.writeToFile(privKey, resp.SignedCertChain)
+			glog.Info("Obtained new cert. Will renew in %s", waittime.String())
+			retries = 0
+			retrialInterval = CertRequestRetrialInterval
+			<-timer.C
+		} else {
+			if retries >= CertRequestMaxRetries {
+				glog.Fatalf("Max number of retrials has been reached: %d. Abort.", CertRequestMaxRetries)
```

Comment:
How about 'Node agent can't get the CSR approved from Istio CA after max number of retrials, please fix the error and retry later.'